backends/oauth2.go

```diff
@@ -0,0 +1,361 @@
+		return false, nil
+	}
+
+	err := o.updateCache(&cache)
+	if err != nil {
+		return false, err
+	}
+
+	res := o.checkAccessToTopic(topic, acc, &cache, username, clientid)
+	log.Debugf("ACL check was %t", res)
+	return res, nil
+}
+
+func (o *Oauth2) GetName() string {
+	return "OAuth2 backend, version " + o.version
+}
+
+func (o *Oauth2) Halt() {
+	// Do whatever cleanup is needed.
+}
+
+func (o *Oauth2) updateCache(cache *userState) error {
+
+	if o.cacheIsValid(cache) {
+		log.Debugf("using cached userinfo for user '%s' to authorize", cache.username)
+	} else {
+		log.Debugf("update userinfo for user '%s' using user info url %s", cache.username, o.userInfoURL)
+
+		info, err := o.getUserInfo(cache.client)
+
+		if err != nil {
+			log.Errorf("Failed to receive UserInfo for user %s: %s", cache.username, err)
+			return err
+		}
+
+		cache.superuser = info.MQTT.Superuser
+		cache.readTopics = info.MQTT.Topics.Read
+		cache.writeTopics = info.MQTT.Topics.Write
+		cache.subscribeTopics = info.MQTT.Topics.Subscribe
+		cache.denyTopics = info.MQTT.Topics.Deny
+		cache.updatedAt = time.Now()
+	}
+
+	log.Debugf("  user is superuser: %t", cache.superuser)
+	log.Debugf("  topics with read permission: %s", cache.readTopics)
+	log.Debugf("  topics with write permission: %s", cache.writeTopics)
+	log.Debugf("  topics with subscribe permission: %s", cache.subscribeTopics)
+	log.Debugf("  denied topics: %s", cache.denyTopics)
+
+	return nil
+}
+
+func (o *Oauth2) createUserWithClientCredentials(username, password, clientid string) (bool, error) {
+	clientcredentialsConfig := go_clientcredentials.Config{
+		ClientID:     username,
+		ClientSecret: password,
+		TokenURL:     o.tokenUrl,
+		Scopes:       o.scopesSplit,
+	}
+
+	clientcredentialsClient := clientcredentialsConfig.Client(context.Background())
+
+	o.userCache[username] = userState{
+		username:  username,
+		superuser: false,
+		createdAt: time.Now(),
+		updatedAt: time.Unix(0, 0),
+		client:    clientcredentialsClient,
+	}
+
+	cache, _ := o.userCache[username]
+	err := o.updateCache(&cache)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (o *Oauth2) createUserWithToken(accessToken, clientid string) (bool, error) {
+	token := &go_oauth2.Token{
+		AccessToken: accessToken,
+		TokenType:   "Bearer",
+	}
+
+	oauth2Config := go_oauth2.Config{
+		Scopes:      o.scopesSplit,
+		RedirectURL: "",
+		Endpoint: go_oauth2.Endpoint{
+			TokenURL: o.tokenUrl,
+			AuthURL:  "",
+		},
+	}
+
+	client := oauth2Config.Client(context.Background(), token)
+
+	o.userCache[accessToken] = userState{
+		username:  accessToken,
+		createdAt: time.Now(),
+		updatedAt: time.Unix(0, 0),
+		client:    client,
+	}
+
+	cache, _ := o.userCache[accessToken]
+	err := o.updateCache(&cache)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (o *Oauth2) getUserInfo(client *http.Client) (*UserInfo, error) {
+	req, _ := http.NewRequest("GET", o.userInfoURL, nil)
```

Comment:
We could use the exported `http` package request verbs instead of "GET".
Also, why are we bypassing the possible error? An http request can definitely fail!